test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,133 @@
+    def test_summon_full_param_with_writeback(self):
+        model = FSDP(nn.Linear(3, 3).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with torch.no_grad():
+            with model._summon_full_params(volatile=False):
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], 0)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_shard_value(self):
+
+        raw_model = nn.Linear(10, 11)
+        # params: 11x10 + 11 -> 121
+        self.assertEqual(121, sum([m.numel() for m in raw_model.parameters()]))
+
+        model = FSDP(raw_model.cuda(self.rank))
+
+        # params should be half as many
+        expected_shard_size = int(math.ceil(121 / self.world_size))
+        self.assertEqual(expected_shard_size, sum([m.numel() for m in model.parameters()]))
+
+        # we're assuming a single flatenned param
+        self.assertEqual(1, len(list(model.parameters())))
+
+        my_shard = torch.clone(next(model.parameters()))
+
+        with model._summon_full_params():
+            # now we see them all
+            self.assertEqual(121, sum([m.numel() for m in model.parameters()]))
+            all_shards = next(model.parameters())
+
+            # this assumes FSDP will chunk the flatenned tensor
+            slice_start = self.rank * expected_shard_size
+            slice_end = (self.rank + 1) * expected_shard_size
+            my_slice = all_shards[slice_start : slice_end]
+            # shards are padded but the full_param tensor is not
+            if self.rank == self.world_size - 1:
+                self.assertTrue(torch.equal(my_shard[0: my_slice.numel()], my_slice))
+            else:
+                self.assertTrue(torch.equal(my_shard, my_slice))
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_recursive(self):
+        raw_model = nn.Sequential(
+            nn.Linear(5, 6, bias=False),
+            nn.Linear(6, 3, bias=False)
+        )
+
+        # this should wrap the first Linear layer, but not the second
+        my_wrap = functools.partial(default_auto_wrap_policy, min_num_params=20)
+        model = FSDP(raw_model.cuda(self.rank), fsdp_auto_wrap_policy=my_wrap)
+
+        outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+        self.assertEqual(9, outer_param.numel())
+        self.assertEqual(15, inner_param.numel())
+
+        # non recursive, should only summon outer
+        with model._summon_full_params(recurse=False):
+            outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+            inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+            self.assertEqual(18, outer_param.numel())
```

Comment:
Assuming its 18 because 9 is the shard size and world size is 2, can we generalize it to world size > 2? I am not sure if these tests run with world size > 2 but they definitely should if they don't